src/ws/session.c

```diff
@@ -553,10 +552,19 @@ perform_tlscert (const char *rhost)
 }
 
 
-static void
+/* Return path of ccache file (without FILE: prefix), clean this up at session end */
+static const char *
 store_krb_credentials (gss_cred_id_t creds, uid_t uid, gid_t gid)
 {
+  gss_key_value_set_desc store;
+  struct gss_key_value_element_struct element;
+  OM_uint32 major, minor;
+  static char ccache[PATH_MAX];
+  int res;
+
   assert (creds != GSS_C_NO_CREDENTIAL);
+  /* can only be called once */
+  assert (ccache[0] == '\0');
```

Comment:
Why not have an actual if check? assert calls are eliminated in optimized builds, normally ...

Reply:
Calling it twice would be an internal programming error, so I thought an assertion would be appropriate (we do run tests with a devel build as well, and even our upstream production builds don't compile out assertions). It's similar to the creds assertion above it. But I don't mind much to turn these into runtime checks. I'll wait for @allisonkarlitskaya 's review (I'm sure she finds some other stuff) and can then update this.

Thanks for your review!

Reply:
My opinion is that we shy away from memory allocation too much when we write C.

Just allocate a string?  If you don't like it because you can't return it and later pass it to `unlink()` then return the entire string and add a new trivial function for `delete_krb_credentials()` and do the check for `FILE:` and `+5` there instead, and bonus `free()`?

Bonus is you can use our `asprintfx()` helper and not argue about if you need to add or remove one to the length :)

Reply:
You are right that in general I am a bit allergic against dynamic allocations, and doubly so in suid root stuff. But in this case the assertion is not just a crutch for the static memory, but an actual first-class intended thing. We should *never* store credentials twice, as the second one would clobber the first, and that'd be an internal program logic error.

If you don't like that style, I'm also okay with having a `static char *ccache` and ensuring that it's `NULL`, and using asprintfx().

Reply:
In general I'm also hoping to dramatically reduce the use of globals and static locals.  I have a hard time imagining a scenario under which we accidentally call this function twice.

If you insist, though, I accept the compromise. :)